ttax/ops.py

```diff
@@ -1,27 +1,82 @@
-      vmapped = func
-      for _ in range(tt_arg.num_batch_dims):
-        vmapped = jax.vmap(vmapped)
-      return vmapped(*args, **kwargs)
-  return vectorized_func
+def tt_vmap(num_batch_args=None):
+  """Decorator which makes a function support batch TT-inputs.
+  Arg:
+    num_batch_args - integer or None.
+      If None, than function will be mapped over all arguments.
+      If integer, specifies the count of first arguments to map over, e.g.
+      num_batch_args=n means that function will be mapped over 
+      first n arguments.
+  Returns:
+    Decorator
+  Comments:
+    The function is vmapped num_batch_dims times, as it supports 
+    multidimensional batches. The number of batch dimension to be 
+    mapped over is shown by num_batch_dims property and should be 
+    the same for all args of the function, by which it will be mapped over. 
+    Otherwise such axis should be specified by num_batch_args.
+    Broadcasting case (when several axis of batches equal to 1) is
+    supported.
+  """
+  def tt_vmap_fixed_batching_pattern(func):
+    @functools.wraps(func)
+    def vectorized_func(*args, **kwargs):
+      tt_arg = args[0]  # TODO: what if only kwargs are present?
+      if tt_arg.num_batch_dims == 0:
+        return func(*args, **kwargs)
+      else:
+        if num_batch_args is not None:
+          num_non_batch_args = len(args) + len(kwargs) - num_batch_args
+          in_axis = [0] * num_batch_args + [None] * num_non_batch_args
+        else:
+          in_axis = [0] * (len(args) + len(kwargs))
+        broadcasting_case = False
+        if len(in_axis) > 1 and (isinstance(args[1], TTMatrix) or
+                                 isinstance(args[1], TT)):
+          if args[0].is_tt_matrix != args[1].is_tt_matrix:
+            raise ValueError('Types of the arguments are different.')
+          if not are_batches_broadcastable(args[0], args[1]):
+            raise ValueError('The batch sizes are different and not 1, '
+                             'broadcasting is not available.')
+          broadcast_shape = np.maximum(list(args[0].batch_shape), 
+                                       list(args[1].batch_shape))
+          new_args = []
+          broadcasting_case = True
+          if args[0].is_tt_matrix:
+            for tt in args[:2]:
+              new_cores = []
+              for core in tt.tt_cores:
+                core = jnp.broadcast_to(core, list(broadcast_shape) +
+                                              list(core.shape[-4:]))
+                new_cores.append(core)
+              new_args.append(TTMatrix(new_cores))
+          else:
+            for tt in args[:2]:
+              new_cores = []
+              for core in tt.tt_cores:
+                core = jnp.broadcast_to(core, list(broadcast_shape) + 
+                                              list(core.shape[-3:]))
+                new_cores.append(core)
+              new_args.append(TT(new_cores)) 
+        vmapped = func
```

Comment:
You can make

```
else:
  new_args = args
```

to avoid `if broadcasting_case:` below (and also avoid introducing `broadcasting_case` variable)